bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/directentrylogger/DirectWriter.java

```diff
@@ -0,0 +1,322 @@
+        synchronized (outstandingWrites) {
+            Iterator<Future<?>> iter = outstandingWrites.iterator();
+            while (iter.hasNext()) { // clear out completed futures
+                Future<?> f = iter.next();
+                waitForFuture(f);
+                iter.remove();
+            }
+        }
+    }
+
+    private void waitForFuture(Future<?> f) throws IOException {
+        try {
+            f.get();
+        } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw new IOException(ie);
+        } catch (Throwable t) {
+            if (t.getCause() instanceof IOException) {
+                throw (IOException) t.getCause();
+            } else {
+                throw new IOException(t);
+            }
+        }
+    }
+
+    private void flushBuffer() throws IOException {
+        synchronized (bufferLock) {
+            if (this.nativeBuffer != null) {
+                int bytesToWrite = this.nativeBuffer.padToAlignment();
+                if (bytesToWrite == 0) {
+                    return;
+                }
+
+                Buffer bufferToFlush = this.nativeBuffer;
+                this.nativeBuffer = null;
+
+                long offsetToWrite = offset;
+                offset += bytesToWrite;
+
+                Future<?> f = writeExecutor.submit(() -> {
+                        try {
+                            if (bytesToWrite <= 0) {
+                                return null;
+                            }
+                            int ret = nativeIO.pwrite(fd, bufferToFlush.pointer(), bytesToWrite, offsetToWrite);
+                            if (ret != bytesToWrite) {
+                                throw new IOException(exMsg("Incomplete write")
+                                                      .kv("filename", filename)
+                                                      .kv("pointer", bufferToFlush.pointer())
+                                                      .kv("offset", offsetToWrite)
+                                                      .kv("writeSize", bytesToWrite)
+                                                      .kv("bytesWritten", ret).toString());
+                            }
+                        } catch (NativeIOException ne) {
+                            throw new IOException(exMsg(ne.getMessage())
+                                                  .kv("filename", filename)
+                                                  .kv("offset", offsetToWrite)
+                                                  .kv("writeSize", bytesToWrite)
+                                                  .kv("pointer", bufferToFlush.pointer())
+                                                  .kv("errno", ne.getErrno()).toString(), ne);
+                        } finally {
+                            bufferPool.release(bufferToFlush);
+                        }
+                        return null;
```

Comment:
The max line length is 112, so no need to refactor.

Reply:
I meant the code at https://github.com/apache/bookkeeper/blob/44f0f1ae9758fdfd8b9e3687b786a5f595c768d6/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/directentrylogger/DirectWriter.java#L118-L138 looks similar, can probably refactor, this will also guarantee "if (bytesToWrite <= 0)" check in both places

Reply:
done